<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Converting single-cell data structures between Bioconductor and Python â€¢ zellkonverter</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Converting single-cell data structures between Bioconductor and Python">
<meta property="og:description" content="zellkonverter">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">zellkonverter</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/zellkonverter.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/theislab/zellkonverter/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="zellkonverter_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Converting single-cell data structures between Bioconductor and Python</h1>
                        <h4 class="author">Luke Zappia</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:luke@lazappi.id.au" class="email">luke@lazappi.id.au</a>
      
                              <h4 class="author">Aaron Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 class="date">Revised: 19 June 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/theislab/zellkonverter/blob/master/vignettes/zellkonverter.Rmd"><code>vignettes/zellkonverter.Rmd</code></a></small>
      <div class="hidden name"><code>zellkonverter.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>This package provides a lightweight interface between the Bioconductor <code>SingleCellExperiment</code> data structure and the Python <code>AnnData</code>-based single-cell analysis environment. The idea is to enable users and developers to easily move data between these frameworks to construct a multi-language analysis pipeline across R/Bioconductor and Python.</p>
</div>
<div id="reading-and-writing-h5ad-files" class="section level1">
<h1 class="hasAnchor">
<a href="#reading-and-writing-h5ad-files" class="anchor"></a>Reading and writing H5AD files</h1>
<p>The <code><a href="../reference/readH5AD.html">readH5AD()</a></code> function can be used to read a <code>SingleCellExperiment</code> from a H5AD file. This can be manipulated in the usual way as described in the <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment">SingleCellExperiment</a></em> documentation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/theislab/zellkonverter">zellkonverter</a></span><span class="op">)</span>

<span class="co"># Obtaining an example H5AD file.</span>
<span class="va">example_h5ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"krumsiek11.h5ad"</span>,
                            package <span class="op">=</span> <span class="st">"zellkonverter"</span><span class="op">)</span>
<span class="fu"><a href="../reference/readH5AD.html">readH5AD</a></span><span class="op">(</span><span class="va">example_h5ad</span><span class="op">)</span></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 11 640 
## metadata(2): highlights iroot
## assays(1): X
## rownames(11): Gata2 Gata1 ... EgrNab Gfi1
## rowData names(0):
## colnames(640): 0 1 ... 158-3 159-3
## colData names(1): cell_type
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<p>We can also write a <code>SingleCellExperiment</code> to a H5AD file with the <code><a href="../reference/writeH5AD.html">writeH5AD()</a></code> function. This is demonstrated below on the classic Zeisel mouse brain dataset from the <em><a href="https://bioconductor.org/packages/3.12/scRNAseq">scRNAseq</a></em> package. The resulting file can then be directly used in compatible Python-based analysis frameworks.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span>

<span class="va">sce_zeisel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ZeiselBrainData.html">ZeiselBrainData</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">out_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".h5ad"</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeH5AD.html">writeH5AD</a></span><span class="op">(</span><span class="va">sce_zeisel</span>, file <span class="op">=</span> <span class="va">out_path</span><span class="op">)</span></pre></div>
</div>
<div id="converting-between-singlecellexperiment-and-anndata-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#converting-between-singlecellexperiment-and-anndata-objects" class="anchor"></a>Converting between <code>SingleCellExperiment</code> and <code>AnnData</code> objects</h1>
<p>Developers and power users who control their Python environments can directly convert between <code>SingleCellExperiment</code> and <a href="https://anndata.readthedocs.io/en/stable/"><code>AnnData</code> objects</a> using the <code><a href="../reference/AnnData-Conversion.html">SCE2AnnData()</a></code> and <code><a href="../reference/AnnData-Conversion.html">AnnData2SCE()</a></code> utilities. These functions expect that <em><a href="https://CRAN.R-project.org/package=reticulate">reticulate</a></em> has already been loaded along with an appropriate version of the <a href="https://pypi.org/project/anndata/"><em>anndata</em></a> package. We suggest using the <em><a href="https://bioconductor.org/packages/3.12/basilisk">basilisk</a></em> package to set up the Python environment before using these functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">basilisk</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span>

<span class="va">seger</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/SegerstolpePancreasData.html">SegerstolpePancreasData</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">roundtrip</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/basilisk/man/basiliskStart.html">basiliskRun</a></span><span class="op">(</span>fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">{</span>
     <span class="co"># Convert SCE to AnnData:</span>
     <span class="va">adata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AnnData-Conversion.html">SCE2AnnData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

     <span class="co"># Maybe do some work in Python on 'adata':</span>
     <span class="co"># BLAH BLAH BLAH</span>

     <span class="co"># Convert back to an SCE:</span>
     <span class="fu"><a href="../reference/AnnData-Conversion.html">AnnData2SCE</a></span><span class="op">(</span><span class="va">adata</span><span class="op">)</span>
<span class="op">}</span>, env <span class="op">=</span> <span class="fu">zellkonverter</span><span class="fu">:::</span><span class="va">anndata_env</span>, sce <span class="op">=</span> <span class="va">seger</span><span class="op">)</span></pre></div>
<p>Package developers can guarantee that they are using the same versions of Python packages as <em><a href="https://bioconductor.org/packages/3.12/zellkonverter">zellkonverter</a></em> by using the <code>.AnnDataDependencies</code> variable to set up their Python environments.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">.AnnDataDependencies</span></pre></div>
<pre><code>## [1] "anndata==0.7.4"  "h5py==2.10.0"    "hdf5==1.10.5"    "natsort==7.0.1" 
## [5] "numpy==1.19.1"   "packaging==20.4" "pandas==1.1.2"   "scipy==1.5.2"   
## [9] "sqlite==3.33.0"</code></pre>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<pre><code>## R version 4.0.3 Patched (2020-10-21 r79360)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel stats    graphics utils    stats4   methods  base    
## 
## other attached packages:
##  [1] basilisk_1.2.0              scRNAseq_2.3.17            
##  [3] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
##  [5] Biobase_2.50.0              GenomicRanges_1.42.0       
##  [7] GenomeInfoDb_1.26.0         IRanges_2.24.0             
##  [9] S4Vectors_0.28.0            BiocGenerics_0.36.0        
## [11] MatrixGenerics_1.2.0        matrixStats_0.57.0         
## [13] zellkonverter_1.1.0         knitr_1.30                 
## [15] BiocStyle_2.18.0           
## 
## loaded via a namespace (and not attached):
##  [1] ProtGenerics_1.22.0           bitops_1.0-6                 
##  [3] fs_1.5.0                      bit64_4.0.5                  
##  [5] filelock_1.0.2                progress_1.2.2               
##  [7] httr_1.4.2                    rprojroot_1.3-2              
##  [9] tools_4.0.3                   backports_1.1.10             
## [11] R6_2.4.1                      lazyeval_0.2.2               
## [13] DBI_1.1.0                     prettyunits_1.1.1            
## [15] tidyselect_1.1.0              bit_4.0.4                    
## [17] curl_4.3                      compiler_4.0.3               
## [19] textshaping_0.1.2             datasets_4.0.3               
## [21] basilisk.utils_1.2.0          xml2_1.3.2                   
## [23] desc_1.2.0                    DelayedArray_0.16.0          
## [25] rtracklayer_1.50.0            bookdown_0.20                
## [27] askpass_1.1                   rappdirs_0.3.1               
## [29] pkgdown_1.6.1.9000            Rsamtools_2.6.0              
## [31] systemfonts_0.3.2             stringr_1.4.0                
## [33] digest_0.6.25                 rmarkdown_2.4                
## [35] XVector_0.30.0                pkgconfig_2.0.3              
## [37] htmltools_0.5.0               ensembldb_2.14.0             
## [39] dbplyr_1.4.4                  fastmap_1.0.1                
## [41] grDevices_4.0.3               rlang_0.4.8                  
## [43] RSQLite_2.2.1                 shiny_1.5.0                  
## [45] generics_0.0.2                jsonlite_1.7.1               
## [47] BiocParallel_1.24.0           dplyr_1.0.2                  
## [49] RCurl_1.98-1.2                magrittr_1.5                 
## [51] GenomeInfoDbData_1.2.4        Matrix_1.2-18                
## [53] Rcpp_1.0.5                    reticulate_1.16              
## [55] lifecycle_0.2.0               stringi_1.5.3                
## [57] yaml_2.2.1                    zlibbioc_1.36.0              
## [59] BiocFileCache_1.14.0          AnnotationHub_2.22.0         
## [61] grid_4.0.3                    blob_1.2.1                   
## [63] promises_1.1.1                ExperimentHub_1.16.0         
## [65] crayon_1.3.4                  lattice_0.20-41              
## [67] Biostrings_2.58.0             GenomicFeatures_1.42.0       
## [69] hms_0.5.3                     pillar_1.4.6                 
## [71] biomaRt_2.46.0                XML_3.99-0.5                 
## [73] glue_1.4.2                    BiocVersion_3.12.0           
## [75] evaluate_0.14                 BiocManager_1.30.10          
## [77] vctrs_0.3.4                   httpuv_1.5.4                 
## [79] openssl_1.4.3                 purrr_0.3.4                  
## [81] assertthat_0.2.1              xfun_0.18                    
## [83] mime_0.9                      xtable_1.8-4                 
## [85] AnnotationFilter_1.14.0       later_1.1.0.1                
## [87] ragg_0.4.0                    tibble_3.0.4                 
## [89] GenomicAlignments_1.26.0      AnnotationDbi_1.52.0         
## [91] memoise_1.1.0                 ellipsis_0.3.1               
## [93] interactiveDisplayBase_1.28.0</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Luke Zappia, Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
